2022/Q3/artifacts/PoS-pseudocode/PoS-model.md

```diff
@@ -188,19 +196,59 @@ tx_delegate(validator_address, delegator_address, amount)
 ```go
 tx_undelegate(validator_address, delegator_address, amount)
 {
-  unbond(validator_address, delegator_address, amount)
+  unbond(validator_address, delegator_address, amount, true)
 }
 ```
 
-<!-- 
 ```go
-tx_redelegate(src_validator_address, dest_validator_address, delegator_address, amount)
+// Amount is not specified - a re-delegation must transfer the full amount
+tx_redelegate(src_validator_address, dest_validator_address, delegator_address)
 {
-  unbond(src_validator_address, delegator_address, amount)
-  bond(dst_validator_address, delegator_address, amount, unbonding_length)
+
+  // Disallow re-delegation if either validator is frozen (similar to unbonding)
+  var src_frozen = read_epoched_field(validators[src_validator_address].frozen, cur_epoch, false)
+  var dest_frozen = read_epoched_field(validators[dest_validator_address].frozen, cur_epoch, false)
+  if (is_validator(src_validator_address, cur_epoch+pipeline_length) && src_frozen == false &&
+      is_validator(dest_validator_address, cur_epoch+pipeline_length) && dest_frozen == false
+  ) then
+    // Check that `incoming_redelegations[delegator_address]` for `dest_validator_address` and 
+    // `outgoing_redelegations[delegator_address]` for `src_validator_address` either don't exist
+    // or if they do, they cannot be slashed anymore (`end + unbonding_length <= cur_epoch`)
+    var incoming = validators[dest_validator_address].incoming_redelegations[src_validator_address][delegator_address]
+    var outgoing = validators[src_validator_address].outgoing_redelegations[dest_validator_address][delegator_address]
+    if ((incoming != ⊥ && incoming.end + unbonding_length > cur_epoch) ||
+       (outgoing != ⊥ && outgoing.end + unbonding_length > cur_epoch)) then
+      return
+
+    // Find the sum of bonded tokens to `src_validator_address`
+    var bonded_tokens = 0
+    var delbonds = {<start, amount> | amount = bonds[delegator_address][src_validator_address].deltas[(start, ⊥)] > 0 && start <= cur_epoch + unbonding_length}
+    forall (<start,amount> in delbonds) do
+      // Apply slashes (and rewards - but we're not dealing with these here)
+      var amount_after_slashing = amount
+      forall (slash in slashes[validator_address] s.t. start <= slash.epoch)
+        amount_after_slashing -= amount*slash.rate
+      bonded_tokens += amount_after_slashing
+
+    // Unbond the tokens from `src_validator_address` at pipeline offset, but 
+    // without recording it in the `unbonds[delegator_address][src_validator_address]`
+    var record_unbonds = false
+    unbond(src_validator_address, delegator_address, bonded_tokens, record_unbonds)
```

Comment:
Good question! I think we're applying slashes on the amount that's being used to update for `update_total_deltas`, but not on the `unbonds` record (this is only set when the newly added `record_unbonds` bool here is true - from this call it's false). This is because the validator total deltas are slashed as soon as the slash is processed, but bonds are only slashed once withdrawn